src/panic.rs

```diff
@@ -16,15 +17,46 @@ Python interpreter to exit.
 );
 
 impl PanicException {
-    // Try to format the error in the same way panic does
+    /// Creates a new PanicException from a panic payload.
+    ///
+    /// Attempts to format the error in the same way panic does.
+    ///
+    /// If dropping `payload` panics, will fatally abort.
     #[cold]
     pub(crate) fn from_panic_payload(payload: Box<dyn Any + Send + 'static>) -> PyErr {
-        if let Some(string) = payload.downcast_ref::<String>() {
-            Self::new_err((string.clone(),))
-        } else if let Some(s) = payload.downcast_ref::<&str>() {
-            Self::new_err((s.to_string(),))
+        let err = if let Some(msg) = panic_payload_to_string(payload) {
+            Self::new_err((msg,))
         } else {
             Self::new_err(("panic from Rust code",))
-        }
+        };
+        err
     }
 }
+
+/// Formats a panic payload into a string, and drops the panic payload.
+///
+/// If dropping the panic payload panics, aborts the process.
+pub(crate) fn panic_payload_to_string(payload: Box<dyn Any + Send + 'static>) -> Option<String> {
+    let msg = if let Some(string) = payload.downcast_ref::<String>() {
+        Some(string.clone())
+    } else if let Some(s) = payload.downcast_ref::<&str>() {
+        Some(s.to_string())
+    } else {
+        None
+    };
+    abort_on_unwind(move || drop(payload), "panic while dropping panic payload");
+    msg
+}
+
+/// Executes `f()`. If it panics, attempts to write `msg` to stderr and then aborts.
+pub(crate) fn abort_on_unwind<F: FnOnce()>(f: F, msg: &str) {
+    // Always fine to use AssertUnwindSafe here because a panic will abort.
+    std::panic::catch_unwind(std::panic::AssertUnwindSafe(f))
+        .unwrap_or_else(|_| abort_with_msg(msg));
```

Comment:
Since `unwrap_or_else` is called for the closure's side effect, I think it would be clearer to match this like

```rust
if let Err(_err) = std::panic::catch_unwind(std::panic::AssertUnwindSafe(f)) {
  abort_with_msg(msg);
}
```

I would also feel better if the binding was `_err` instead of `_` to ensure the panic payload (the one caused by `drop(payload)`) is not dropped before calling `abort_with_msg`.

Reply:
(Just as an idea, could we put a type on the stack which panics on `Drop` itself to trigger abort due to a panic during unwind before calling `f` instead of using `catch_unwind` and having to juggle its potential (second order) panic bomp? Maybe already in `panic_payload_to_string`?)

Reply:
👍 this is a great idea! Will change to implement as this.

Reply:
I would prefer an aborting guard as well.